### PR TITLE
Add printing support

### PR DESCRIPTION
**Pull request by diegorondini**
While CUPS itself is part of openembedded-core, most of real world printing use cases require cups-filters to operate printers.
As it turns out meta-printing hasn't been updated for some time. While proposing these patches to meta-printing is an option, I think it makes sense to support (basic) printing use cases in meta-openembedded directly.

### CODE REVIEW

File: meta-oe/recipes-printing/qpdf/qpdf_9.1.0.bb [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+inherit autotools-brokensep gettext
+
+# disable random file detection for cross-compile
+EXTRA_OECONF = "--without-random \
+                --disable-static \
+                --disable-check-autofiles \
+                "
+
+EXTRA_OEMAKE_class-target = "LIBTOOL=${STAGING_BINDIR_CROSS}/${HOST_SYS}-libtool"
```
Comment (kraj) [inline meta-oe/recipes-printing/qpdf/qpdf_9.1.0.bb]: perhaps you can use `LIBTOOL=${HOST_SYS}-libtool`

File: meta-oe/recipes-printing/qpdf/qpdf_9.1.0.bb [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+S="${WORKDIR}/${PN}-release-${PN}-${PV}"
```
Comment (kraj) [inline meta-oe/recipes-printing/qpdf/qpdf_9.1.0.bb]: Use BPN instead of PN

File: meta-oe/recipes-printing/cups/cups-filters.inc [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+SECTION = "console/utils"
+
+DEPENDS = "cups glib-2.0 glib-2.0-native dbus dbus-glib lcms ghostscript poppler qpdf libpng"
+DEPENDS_class-native = "poppler-native glib-2.0-native dbus-native pkgconfig-native gettext-native libpng-native"
+
+SRC_URI = "http://openprinting.org/download/cups-filters/cups-filters-${PV}.tar.gz"
+
+inherit autotools-brokensep gettext pkgconfig
+
+EXTRA_OECONF_append = " --enable-ghostscript --disable-ldap \
```
Comment (kraj) [inline meta-oe/recipes-printing/cups/cups-filters.inc]: this should be EXTRA_OECONF += ".."

File: meta-oe/recipes-printing/cups/cups-filters.inc [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+                       --localstatedir=${localstatedir}/var \
+                       --with-rcdir=no \
+                       --without-php"
+
+EXTRA_OECONF_append_class-native = " --with-pdftops=pdftops \
+                                    --disable-avahi --disable-ghostscript \
+                                    --disable-ldap \
+                                    --with-png --without-jpeg --without-tiff"
+
+EXTRA_OEMAKE = "-j1"
```
Comment (kraj) [inline meta-oe/recipes-printing/cups/cups-filters.inc]: why is this needed ? in case you want to disable parallelism then use `PARALLEL_MAKE = ""`

File: meta-oe/recipes-printing/cups/cups-filters.inc [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+        /usr/share/cups/banners/confidential \
+        /usr/share/cups/banners/unclassified \
+        /usr/share/cups/banners/form \
+        /usr/share/cups/banners/classified \
+        /usr/share/cups/banners/standard \
+"
+
+do_install_append() {
+	# remove banners, braille dirs
+	rm -rf ${D}/usr/share/cups/{banners,braille}
```
Comment (kraj) [inline meta-oe/recipes-printing/cups/cups-filters.inc]: Use `${datadir}` instead of hardcoding /usr/share

File: meta-oe/recipes-printing/qpdf/qpdf_9.1.0.bb (hunk 1 above)
Comment (diegorondini) [reply meta-oe/recipes-printing/qpdf/qpdf_9.1.0.bb]: Changed as requested.

File: meta-oe/recipes-printing/qpdf/qpdf_9.1.0.bb (hunk 2 above)
Comment (diegorondini) [reply meta-oe/recipes-printing/qpdf/qpdf_9.1.0.bb]: Done.

File: meta-oe/recipes-printing/cups/cups-filters.inc (hunk 3 above)
Comment (diegorondini) [reply meta-oe/recipes-printing/cups/cups-filters.inc]: Done.

File: meta-oe/recipes-printing/cups/cups-filters.inc (hunk 4 above)
Comment (diegorondini) [reply meta-oe/recipes-printing/cups/cups-filters.inc]: Don't seem to be needed anymore, removed..
Built successfully with 2 cores.

File: meta-oe/recipes-printing/cups/cups-filters.inc (hunk 5 above)
Comment (diegorondini) [reply meta-oe/recipes-printing/cups/cups-filters.inc]: Done.

**Comment by diegorondini** [comment]
I've pushed an updated version with the requested changes.
If it looks good I'll do some print tests.

File: meta-oe/recipes-printing/qpdf/qpdf_9.1.0.bb [hunk 6]
```diff
@@ -0,0 +1,33 @@
+DESCRIPTION = "PDF transformation/inspection software"
+HOMEPAGE = "http://qpdf.sourceforge.net"
+LICENSE = "Artistic-2.0"
+SECTION = "libs"
+DEPENDS = "libpcre zlib libjpeg-turbo"
+
+SRC_URI = "https://github.com/qpdf/qpdf/archive/release-qpdf-${PV}.tar.gz"
```
Comment (kraj) [inline meta-oe/recipes-printing/qpdf/qpdf_9.1.0.bb]: I see 
`SRC_URI uses unstable GitHub archives [src-uri-bad]`
perhaps its better to use 
```
SRCREV = "some sha"
SRC_URI = "git://github.com/qpdf/qpdf;protocol=https"
S = "${WORKDIR}/git"
```

Comment (diegorondini) [reply meta-oe/recipes-printing/qpdf/qpdf_9.1.0.bb]: Ok, there are tarballs at sourceforge, I will use those.

**Comment by diegorondini** [comment]
Changed fetch URL to use sourceforge and rebased on meta-oe master.

**Comment by diegorondini** [comment]
@kraj I've seen this is in master-next. If you can please don't merge in master yet, as I just noticed cups in oe-core now installs files in `/usr/libexec` rather than `/usr/lib`

**Comment by diegorondini** [comment]
I've tracked down the issue to the cups recipe in oe-core. The cups-config crossscript reports the wrong install path for filters, backends, etc.
https://patchwork.openembedded.org/patch/168719/
I'll update this patch accordingly when the fix enters oe-core.

**Comment by kraj** [comment]
this is now in master

**Comment by diegorondini** [comment]
Thank you. I will post a separate patch to fix the `libdir` vs `libexecdir` issue next week.
